src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,260 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.UseParameterNullChecking
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpUseParameterNullCheckingDiagnosticAnalyzer
+        : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private const string ArgumentNullExceptionName = $"{nameof(System)}.{nameof(ArgumentNullException)}";
+
+        public CSharpUseParameterNullCheckingDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UseParameterNullCheckingId,
+                   EnforceOnBuildValues.UseParameterNullChecking,
+                   CodeStyleOptions2.PreferParameterNullChecking,
+                   CSharpAnalyzersResources.Use_parameter_null_checking,
+                   new LocalizableResourceString(nameof(AnalyzersResources.Null_check_can_be_simplified), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterCompilationStartAction(context =>
+            {
+                var compilation = (CSharpCompilation)context.Compilation;
+                if (compilation.LanguageVersion < LanguageVersionExtensions.CSharpNext)
+                {
+                    return;
+                }
+
+                var argumentNullException = compilation.GetTypeByMetadataName(ArgumentNullExceptionName);
+                if (argumentNullException is null)
+                {
+                    return;
+                }
+
+                var argumentNullExceptionConstructor = argumentNullException.InstanceConstructors.FirstOrDefault(m =>
+                    m.DeclaredAccessibility == Accessibility.Public
+                    && m.Parameters.Length == 1
+                    && m.Parameters[0].Type.SpecialType == SpecialType.System_String);
+                if (argumentNullExceptionConstructor is null)
+                {
+                    return;
+                }
+
+                var objectType = compilation.GetSpecialType(SpecialType.System_Object);
+                var referenceEqualsMethod = (IMethodSymbol?)objectType
+                    .GetMembers(nameof(ReferenceEquals))
+                    .FirstOrDefault(m => m is IMethodSymbol { DeclaredAccessibility: Accessibility.Public, Parameters.Length: 2 });
+
+                context.RegisterSyntaxNodeAction(
+                    context => AnalyzeSyntax(context, argumentNullExceptionConstructor, referenceEqualsMethod),
+                    SyntaxKind.ConstructorDeclaration,
+                    SyntaxKind.MethodDeclaration,
+                    SyntaxKind.LocalFunctionStatement,
+                    SyntaxKind.SimpleLambdaExpression,
+                    SyntaxKind.ParenthesizedLambdaExpression,
+                    SyntaxKind.AnonymousMethodExpression);
```

Comment:
i think constructor/methods are likely fine.  but you could also extend to anything else that takes parameters (indexers/operators/etc.).  If you dont' want to handle that, please put a note explicitly stating that they're in a "don't care" category, as opposed to an unintentional omission.

Reply:
I'll add a note about indexers. I don't want to support it because of the complexity of potentially eliminating statements from both accessors. Operators should be supported, I'll add a test. I'll make a pass over the method kinds and make sure things are covered.